tests/sagemaker/test_sagemaker_plugin.py

```diff
@@ -0,0 +1,615 @@
+    ]
+    assert any([model_name in object_name for object_name in object_names])
+    assert any(
+        [
+            name in config["EndpointConfigName"]
+            for config in sagemaker_client.list_endpoint_configs()["EndpointConfigs"]
+        ]
+    )
+    assert name in [
+        endpoint["EndpointName"] for endpoint in sagemaker_client.list_endpoints()["Endpoints"]
+    ]
+
+
+@pytest.mark.large
+@mock_sagemaker_aws_services
+def test_create_deployment_with_preexisting_name_throws_exception(
+    pretrained_model, sagemaker_plugin
+):
+    name = "test-app"
+    sagemaker_plugin.create_deployment(
+        name=name, model_uri=pretrained_model.model_uri,
+    )
+
+    with pytest.raises(MlflowException) as exc:
+        sagemaker_plugin.create_deployment(
+            name=name, model_uri=pretrained_model.model_uri,
+        )
+
+    assert "an application with the same name already exists" in exc.value.message
+    assert exc.value.error_code == ErrorCode.Name(INVALID_PARAMETER_VALUE)
+
+
+@pytest.mark.large
+@mock_sagemaker_aws_services
+def test_create_deployment_in_synchronous_mode_waits_for_endpoint_creation_to_complete_before_returning(
+    pretrained_model, sagemaker_client, sagemaker_plugin
+):
+    endpoint_creation_latency = 10
+    get_sagemaker_backend(sagemaker_client.meta.region_name).set_endpoint_update_latency(
+        endpoint_creation_latency
+    )
+
+    name = "test-app"
+    deployment_start_time = time.time()
+    sagemaker_plugin.create_deployment(
+        name=name, model_uri=pretrained_model.model_uri, config=dict(synchronous=True),
+    )
+    deployment_end_time = time.time()
+
+    assert (deployment_end_time - deployment_start_time) >= endpoint_creation_latency
+    endpoint_description = sagemaker_client.describe_endpoint(EndpointName=name)
+    assert endpoint_description["EndpointStatus"] == Endpoint.STATUS_IN_SERVICE
+
+
+@pytest.mark.large
+@mock_sagemaker_aws_services
+def test_create_deployment_in_asynchronous_mode_returns_before_endpoint_creation_completes(
+    pretrained_model, sagemaker_client, sagemaker_plugin
+):
+    endpoint_creation_latency = 10
+    get_sagemaker_backend(sagemaker_client.meta.region_name).set_endpoint_update_latency(
+        endpoint_creation_latency
+    )
+
+    name = "test-app"
+    deployment_start_time = time.time()
+    sagemaker_plugin.create_deployment(
+        name=name,
+        model_uri=pretrained_model.model_uri,
+        config=dict(synchronous=False, archive=True),
+    )
+    deployment_end_time = time.time()
+
+    assert (deployment_end_time - deployment_start_time) < endpoint_creation_latency
+    endpoint_description = sagemaker_client.describe_endpoint(EndpointName=name)
+    assert endpoint_description["EndpointStatus"] == Endpoint.STATUS_CREATING
+
+
+# @pytest.mark.large
+# @mock_sagemaker_aws_services
+# def test_deploy_replace_in_asynchronous_mode_returns_before_endpoint_creation_completes(
```

Comment:
Some tests are commented out in this file because they are testing the ability to update a deployment. This functionality will be handled by `update_deployment()` in a follow-up PR.